### PR TITLE
Migrate the VM hash function to Rescue Prime Optimized

### PR DESCRIPTION
**Pull request by grjte**
This addresses #474.

- [x] replace the Winterfell Rescue Prime hash function with RPO from crypto crate
- [x] update initialization of state in hasher core to use new padding rule
- [x] update `rphash` instruction to follow new padding rule initialization
- [x] change absorption of elements in hasher core to overwrite mode
- [x] update constraints (e.g. rewrite of `enforce_rescue_rounds`)
- [x] update `rphash` test, chiplet bus tests, span block tests
- [x] update docs
- [x] update commit log

Note: `miden-core` still depends on `winter-crypto`, but only for for `RandomCoin` and `RandomCoinError`. Should we move these to `miden-crypto` re-exports?

### CODE REVIEW

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks good! Thank you! I left a couple of minor questions/comments inline.

One thing that I think is missing is updates to `MSTREAM` and `PIPE` operations. I think they are are still adding values to the values on the top of the stack instead of replacing them. So, a combination of `MSTREAM` and `RPPERM` operation should now be producing incorrect results - but we are probably not testing this specific case anywhere.

As a part of this work, we should probably rename assembly instructions (inclusive discussion on this is in #590) - but this could be done in a subsequent PR as well.

File: core/Cargo.toml [hunk 1]
```diff
@@ -17,9 +17,10 @@ doctest = false
...
 [features]
 default = ["std"]
-std = ["math/std", "winter-utils/std", "crypto/std"]
+std = ["math/std", "winter-utils/std", "winter-crypto/std"]
 
 [dependencies]
-crypto = { package = "winter-crypto", version = "0.4.2", default-features = false }
 math = { package = "winter-math", version = "0.4.2", default-features = false }
+crypto = { package = "miden-crypto", version = "0.1.0" }
+winter-crypto = { package = "winter-crypto", version = "0.4.2", default-features = false }
```
[the remark this answers is not in the document]
Comment (grjte) [reply core/Cargo.toml:25]: I left a note in the PR description about this:

Note: miden-core still depends on winter-crypto, but only for for `RandomCoin` and `RandomCoinError`. Should we move these to miden-crypto re-exports?

File: docs/src/design/chiplets/hasher.md [hunk 2]
```diff
@@ -12,7 +12,7 @@ The chiplet can be thought of as having a small instruction set of $11$ instruct
 
 | Instruction | Description |
 | ----------- | ----------- |
-| `RPR`       | Executes a single round of Rescue Prime. All cycles which are not one less than a multiple of $8$ execute this instruction. That is, the chiplet executes this instruction on cycles $0, 1, 2, 3, 4, 5, 6$, but not $7$, and then again, $8, 9, 10, 11, 12, 13, 14$, but not $15$ etc.                             |
+| `RPR`       | Executes a single round of Rescue Prime Optimized. All cycles which are not one less than a multiple of $8$ execute this instruction. That is, the chiplet executes this instruction on cycles $0, 1, 2, 3, 4, 5, 6$, but not $7$, and then again, $8, 9, 10, 11, 12, 13, 14$, but not $15$ etc.                             |
```
[the remark this answers is not in the document]
Comment (grjte) [reply docs/src/design/chiplets/hasher.md:15]: Yeah, this is a good idea for reducing future updates.

File: core/src/utils/mod.rs [hunk 3]
```diff
@@ -19,7 +19,8 @@ pub use winter_utils::{
     ByteWriter, Deserializable, DeserializationError, Serializable, SliceReader,
 };
 
-pub use crypto::{RandomCoin, RandomCoinError};
+// TODO: move these to "miden-crypto" package
+pub use winter_crypto::{RandomCoin, RandomCoinError};
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply core/src/utils/mod.rs:23]: Great! Yeah - I think we should re-export them from `miden-crypto`. Once that's done, we can update `miden-vm` in a separate PR.

**Comment by grjte** [comment]
> One thing that I think is missing is updates to `MSTREAM` and `PIPE` operations. I think they are are still adding values to the values on the top of the stack instead of replacing them. So, a combination of `MSTREAM` and `RPPERM` operation should now be producing incorrect results - but we are probably not testing this specific case anywhere.

Good point, thank you! I'll add a test and update those.

> As a part of this work, we should probably rename assembly instructions (inclusive discussion on this is in #590) - but this could be done in a subsequent PR as well.

I agree that we should rename these, but it doesn't seem like there's a consensus yet. We should come to a decision on that in the next few days, but maybe for that reason it should also be done as a subsequent PR. (I don't mind either way though)

**Comment by grjte** [comment]
> a combination of `MSTREAM` and `RPPERM` operation should now be producing incorrect results - but we are probably not testing this specific case anywhere.

I added a new test for combining `adv_pipe` and `mem_stream` to get the same hash result, since we didn't have that before, and I captured the difference of adding vs. not adding the stack elements there.

File: docs/src/user_docs/assembly/io_operations.md [hunk 4]
```diff
@@ -36,7 +36,7 @@ In both case the values must still encode valid field elements.
...
-| adv_pipe <br> - *(2 cycles)*     | [S2, S1, S0, a, ... ] | [T2, T1, T0, b, ... ] | $[T_0, T_1, T_2] \leftarrow permute(S_0, S_1 + tape.next(4), S_2 + tape.next(4))$ <br> $b \leftarrow a + 2$ <br> Removes the next two words (8 elements) from the advice tape, inserts them into memory sequentially starting from address $a$, then adds them to the top 8 elements of the stack and applies a Rescue Prime permutation to the top 12 elements of the stack. At the end of the operation, the address is incremented by $2$. <br> Fails if the advice tape has fewer than $8$ values. |
+| adv_pipe <br> - *(2 cycles)*     | [S2, S1, S0, a, ... ] | [T2, T1, T0, b, ... ] | $[T_0, T_1, T_2] \leftarrow permute(S_0, S_1 + tape.next(4), S_2 + tape.next(4))$ <br> $b \leftarrow a + 2$ <br> Removes the next two words (8 elements) from the advice tape, inserts them into memory sequentially starting from address $a$, then overwrites the top 8 elements of the stack with them, and applies a Rescue Prime Optimized permutation to the top 12 elements of the stack. At the end of the operation, the address is incremented by $2$. <br> Fails if the advice tape has fewer than $8$ values. |
```
Comment (bobbinth) [inline docs/src/user_docs/assembly/io_operations.md:39]: I think the $[T_0, T_1, T_2] \leftarrow permute(S_0, S_1 + tape.next(4), S_2 + tape.next(4))$ formula is incorrect here.

File: docs/src/user_docs/assembly/io_operations.md [hunk 5]
```diff
@@ -50,7 +50,7 @@ Memory is guaranteed to be initialized to zeros. Thus, when reading from memory
...
 | mem_storew <br> - *(1 cycle)*  <br> mem_storew.*a* <br> - *(2-3 cycles)*  | [a, A, ... ] | [A, ... ] | $A \rightarrow mem[a]$ <br> Stores the top four elements of the stack in memory at address $a$. If $a$ is provided via the stack, it is removed from the stack first. <br> Fails if $a \ge 2^{32}$ |
-| mem_stream <br> - *(2 cycles)* | [S2, S1, S0, a, ...] | [T2, T1, T0, b, ...] | $[T_0, T_1, T_2] \leftarrow permute(S_0, S_1 + mem[a], S_2 + mem[a+1])$ <br> $b \leftarrow a + 2$ <br> Loads two words from memory starting at the address $a$, adds them to the top 8 elements of the stack, and applies Rescue Prime permutation to the top 12 elements of the stack. At the end of the operation the address is incremented by $2$. |
+| mem_stream <br> - *(2 cycles)* | [S2, S1, S0, a, ...] | [T2, T1, T0, b, ...] | $[T_0, T_1, T_2] \leftarrow permute(S_0, S_1 + mem[a], S_2 + mem[a+1])$ <br> $b \leftarrow a + 2$ <br> Loads two words from memory starting at the address $a$, overwrites the top 8 elements of the stack with them, and applies Rescue Prime Optimized permutation to the top 12 elements of the stack. At the end of the operation the address is incremented by $2$. |
```
Comment (bobbinth) [inline docs/src/user_docs/assembly/io_operations.md:53]: I think the $[T_0, T_1, T_2] \leftarrow permute(S_0, S_1 + mem[a], S_2 + mem[a+1])$ formula is incorrect here.

**Review by bobbinth — APPROVED** [review]
Looks great! Thank you! I left a couple of comments inline for two doc-related issues. Once these are fixed - it should be good to merge.

**Comment by bobbinth** [comment]
> > As a part of this work, we should probably rename assembly instructions (inclusive discussion on this is in #590) - but this could be done in a subsequent PR as well.
> 
> I agree that we should rename these, but it doesn't seem like there's a consensus yet. We should come to a decision on that in the next few days, but maybe for that reason it should also be done as a subsequent PR. (I don't mind either way though).

Yes - let's do it in a different PR.

File: processor/src/operations/io_ops.rs [hunk 6]
```diff
@@ -387,10 +385,9 @@ mod tests {
         // execute the MSTREAM operation
         process.execute_op(Operation::MStream).unwrap();
 
-        // the result the first 8 values should be the result of adding the values on the stack
-        // to the values in memory (each should result in 35), the next 4 values should remain
+        // the first 8 values should contain the values from memory. the next 4 values should remain
         // unchanged, and the address should be incremented by 2 (i.e., 1 -> 3).
-        let stack_values = [35, 35, 35, 35, 35, 35, 35, 35, 4, 3, 2, 1, 3, 101];
+        let stack_values = [23, 24, 25, 26, 27, 28, 29, 30, 4, 3, 2, 1, 3, 101];
```
Comment (tohrnii) [inline processor/src/operations/io_ops.rs:390]: minor nit: It might be better to write these values in terms of `word1` and `word2` arrays for clarity.

**Review by tohrnii — APPROVED** [review]
Looks good to me. Thanks. I left a couple of non-blocking minor nits.

File: docs/src/user_docs/assembly/io_operations.md (hunk 4 above)
Comment (grjte) [reply docs/src/user_docs/assembly/io_operations.md:39]: Yes, sorry I missed this. Thanks!

File: docs/src/design/chiplets/hasher.md [hunk 7]
```diff
@@ -12,11 +12,11 @@ The chiplet can be thought of as having a small instruction set of $11$ instruct
...
-| `MU`        | Initiates Merkle path verification for the "new" node value during Merkle root update computation. This instruction can be executed only on cycles which are multiples of $8$, and it can also be executed concurrently with `RPR` instruction.                                                                    |
+| `HR`       | Executes a single round of the VM's native hash function. All cycles which are not one less than a multiple of $8$ execute this instruction. That is, the chiplet executes this instruction on cycles $0, 1, 2, 3, 4, 5, 6$, but not $7$, and then again, $8, 9, 10, 11, 12, 13, 14$, but not $15$ etc.                             |
+| `BP`        | Initiates computation of a single permutation, a 2-to-1 hash, or a linear hash of many elements. This instruction can be executed only on cycles which are multiples of $8$, and it can also be executed concurrently with `HR` instruction.                                                                      |
```
[the remark this answers is not in the document]
Comment (grjte) [reply docs/src/design/chiplets/hasher.md:16]: Good catch, thanks. I just did "replace all"

File: processor/src/operations/io_ops.rs (hunk 6 above)
Comment (grjte) [reply processor/src/operations/io_ops.rs:390]: updated!

**Pull request merged**